### ISSUE TITLE
NullPointerException in DataTypeFactory.java:266

### ISSUE THREAD

**Issue by greg0ire** [liquibase/liquibase#2049]
## Environment

**Liquibase Version**: 44.3

**Liquibase Integration & Version**: CLI? Not sure how I can find out, this is running in our CI.

**Liquibase Extension(s) & Version**: liquibase-percona 4.4.3

**Database Vendor & Version**: MySQL

**Operating System Type & Version**: Linux

## Description

There is an exception, and the error handling does say which column (if that's the issue) is problematic.

## Steps To Reproduce

Here is the changelog I am currently trying to apply

```xml
<?xml version="1.0" encoding="UTF-8"?>
<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
                   xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext"
                   xmlns:pro="http://www.liquibase.org/xml/ns/pro" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd http://www.liquibase.org/xml/ns/pro http://www.liquibase.org/xml/ns/pro/liquibase-pro-4.0.xsd http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-4.0.xsd">
    <changeSet id="1"  author="redacted">
        <modifyDataType
            xmlns:liquibasePercona="http://www.liquibase.org/xml/ns/dbchangelog-ext/liquibase-percona"
            liquibasePercona:usePercona="true"
            liquibasePercona:perconaOptions="--force-concat-enums"
            columnName="market"
            newDataType="VARCHAR(10)"
            tableName="redacted"
            />
        <rollback>redacted</rollback>
    </changeSet>
</databaseChangeLog>
```

## Actual Behavior
The following crash happens

```
liquibase.exception.CommandExecutionException: liquibase.exception.LiquibaseException: Unexpected error running Liquibase: Unknown reason
	at liquibase.command.CommandScope.execute(CommandScope.java:153)
	at liquibase.integration.commandline.CommandRunner.call(CommandRunner.java:45)
	at liquibase.integration.commandline.CommandRunner.call(CommandRunner.java:15)
	at picocli.CommandLine.executeUserObject(CommandLine.java:1953)
	at picocli.CommandLine.access$1300(CommandLine.java:145)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2352)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2346)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2311)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
	at picocli.CommandLine.execute(CommandLine.java:2078)
	at liquibase.integration.commandline.LiquibaseCommandLine.lambda$execute$1(LiquibaseCommandLine.java:317)
	at liquibase.Scope.child(Scope.java:175)
	at liquibase.Scope.child(Scope.java:151)
	at liquibase.integration.commandline.LiquibaseCommandLine.execute(LiquibaseCommandLine.java:289)
	at liquibase.integration.commandline.LiquibaseCommandLine.main(LiquibaseCommandLine.java:83)
Caused by: liquibase.exception.LiquibaseException: Unexpected error running Liquibase: Unknown reason
	at liquibase.integration.commandline.Main$1.run(Main.java:447)
	at liquibase.integration.commandline.Main$1.run(Main.java:225)
	at liquibase.Scope.child(Scope.java:175)
	at liquibase.Scope.child(Scope.java:151)
	at liquibase.integration.commandline.Main.run(Main.java:225)
	at liquibase.command.AbstractCliWrapperCommandStep.run(AbstractCliWrapperCommandStep.java:32)
	at liquibase.command.CommandScope.execute(CommandScope.java:147)
	... 14 more
Caused by: liquibase.exception.LiquibaseException: java.lang.NullPointerException
	at liquibase.Liquibase.runInScope(Liquibase.java:2409)
	at liquibase.Liquibase.tag(Liquibase.java:1848)
	at liquibase.integration.commandline.Main.doMigration(Main.java:1726)
	at liquibase.integration.commandline.Main$1.lambda$run$0(Main.java:402)
	at liquibase.Scope.lambda$child$0(Scope.java:166)
	at liquibase.Scope.child(Scope.java:175)
	at liquibase.Scope.child(Scope.java:165)
	at liquibase.Scope.child(Scope.java:144)
	at liquibase.integration.commandline.Main$1.run(Main.java:401)
	... 20 more
Caused by: java.lang.NullPointerException
	at liquibase.datatype.DataTypeFactory.fromObject(DataTypeFactory.java:266)
	at liquibase.sqlgenerator.core.TagDatabaseGenerator.generateSql(TagDatabaseGenerator.java:35)
	at liquibase.sqlgenerator.core.TagDatabaseGenerator.generateSql(TagDatabaseGenerator.java:17)
	at liquibase.sqlgenerator.SqlGeneratorChain.generateSql(SqlGeneratorChain.java:30)
	at liquibase.sqlgenerator.SqlGeneratorFactory.generateSql(SqlGeneratorFactory.java:220)
	at liquibase.executor.AbstractExecutor.applyVisitors(AbstractExecutor.java:105)
	at liquibase.executor.jvm.JdbcExecutor.access$400(JdbcExecutor.java:40)
	at liquibase.executor.jvm.JdbcExecutor$ExecuteStatementCallback.doInStatement(JdbcExecutor.java:372)
	at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:82)
	at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:150)
	at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:134)
	at liquibase.changelog.StandardChangeLogHistoryService.tag(StandardChangeLogHistoryService.java:440)
	at liquibase.database.AbstractJdbcDatabase.tag(AbstractJdbcDatabase.java:877)
	at liquibase.Liquibase$15.run(Liquibase.java:1860)
	at liquibase.Scope.lambda$child$0(Scope.java:166)
	at liquibase.Scope.child(Scope.java:175)
	at liquibase.Scope.child(Scope.java:165)
	at liquibase.Scope.child(Scope.java:144)
	at liquibase.Liquibase.runInScope(Liquibase.java:2404)
	... 28 more
Unexpected error running Liquibase: java.lang.NullPointerException
liquibase.exception.CommandExecutionException: liquibase.exception.LiquibaseException: Unexpected error running Liquibase: Unknown reason
	at liquibase.command.CommandScope.execute(CommandScope.java:153)
	at liquibase.integration.commandline.CommandRunner.call(CommandRunner.java:45)
	at liquibase.integration.commandline.CommandRunner.call(CommandRunner.java:15)
	at picocli.CommandLine.executeUserObject(CommandLine.java:1953)
	at picocli.CommandLine.access$1300(CommandLine.java:145)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2352)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2346)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2311)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
	at picocli.CommandLine.execute(CommandLine.java:2078)
	at liquibase.integration.commandline.LiquibaseCommandLine.lambda$execute$1(LiquibaseCommandLine.java:317)
	at liquibase.Scope.child(Scope.java:175)
	at liquibase.Scope.child(Scope.java:151)
	at liquibase.integration.commandline.LiquibaseCommandLine.execute(LiquibaseCommandLine.java:289)
	at liquibase.integration.commandline.LiquibaseCommandLine.main(LiquibaseCommandLine.java:83)
Caused by: liquibase.exception.LiquibaseException: Unexpected error running Liquibase: Unknown reason
	at liquibase.integration.commandline.Main$1.run(Main.java:447)
	at liquibase.integration.commandline.Main$1.run(Main.java:225)
	at liquibase.Scope.child(Scope.java:175)
	at liquibase.Scope.child(Scope.java:151)
	at liquibase.integration.commandline.Main.run(Main.java:225)
	at liquibase.command.AbstractCliWrapperCommandStep.run(AbstractCliWrapperCommandStep.java:32)
	at liquibase.command.CommandScope.execute(CommandScope.java:147)
	... 14 more
Caused by: liquibase.exception.LiquibaseException: java.lang.NullPointerException
	at liquibase.Liquibase.runInScope(Liquibase.java:2409)
	at liquibase.Liquibase.tag(Liquibase.java:1848)
	at liquibase.integration.commandline.Main.doMigration(Main.java:1726)
	at liquibase.integration.commandline.Main$1.lambda$run$0(Main.java:402)
	at liquibase.Scope.lambda$child$0(Scope.java:166)
	at liquibase.Scope.child(Scope.java:175)
	at liquibase.Scope.child(Scope.java:165)
	at liquibase.Scope.child(Scope.java:144)
	at liquibase.integration.commandline.Main$1.run(Main.java:401)
	... 20 more
Caused by: java.lang.NullPointerException
	at liquibase.datatype.DataTypeFactory.fromObject(DataTypeFactory.java:266)
	at liquibase.sqlgenerator.core.TagDatabaseGenerator.generateSql(TagDatabaseGenerator.java:35)
	at liquibase.sqlgenerator.core.TagDatabaseGenerator.generateSql(TagDatabaseGenerator.java:17)
	at liquibase.sqlgenerator.SqlGeneratorChain.generateSql(SqlGeneratorChain.java:30)
	at liquibase.sqlgenerator.SqlGeneratorFactory.generateSql(SqlGeneratorFactory.java:220)
	at liquibase.executor.AbstractExecutor.applyVisitors(AbstractExecutor.java:105)
	at liquibase.executor.jvm.JdbcExecutor.access$400(JdbcExecutor.java:40)
	at liquibase.executor.jvm.JdbcExecutor$ExecuteStatementCallback.doInStatement(JdbcExecutor.java:372)
	at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:82)
	at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:150)
	at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:134)
	at liquibase.changelog.StandardChangeLogHistoryService.tag(StandardChangeLogHistoryService.java:440)
	at liquibase.database.AbstractJdbcDatabase.tag(AbstractJdbcDatabase.java:877)
	at liquibase.Liquibase$15.run(Liquibase.java:1860)
	at liquibase.Scope.lambda$child$0(Scope.java:166)
	at liquibase.Scope.child(Scope.java:175)
	at liquibase.Scope.child(Scope.java:165)
	at liquibase.Scope.child(Scope.java:144)
	at liquibase.Liquibase.runInScope(Liquibase.java:2404)
	... 28 more
```

I think this is were the issue lies: https://github.com/liquibase/liquibase/blob/dafabf76290d767fc4910bdc5111016b99cbeaf3/liquibase-core/src/main/java/liquibase/datatype/DataTypeFactory.java#L262-L267

## Expected/Desired Behavior
I would expect either no exception, or an exception telling me what needs to be changed in my changelog file or in the database

## Additional Context

This happened after a tentative migration from liquibase 3.10.3 to 4.4.3. and liquibase-percona from 1.7.1 to 4.4.3, not sure if there is an upgrade guide somewhere I should read :grimacing:

**Comment by molivasdat** [comment]
@greg0ire Yes, you are correct in that I could move it over there but it would be better for you to be notified and show that you submitted the issue rather than myself.  You can leave it here for now.  A little lower in the stack trace there is a call to TagDatabaseGenerator.generateSql.  I don't know if there is a special tag or if you know the changeset that the NPE is occurring or another way to recreate this so that it can be fixed.  Even better if you can find it and submit a PR to fix it.

**Comment by greg0ire** [comment]
> Yes, you are correct in that I could move it over there but it would be better for you to be notified and show that you submitted the issue rather than myself

If you don't do that manually, but with the "transfer" button, I think that will be the case.

>  A little lower in the stack trace there is a call to TagDatabaseGenerator.generateSql. I don't know if there is a special tag or if you know the changeset that the NPE is occurring or another way to recreate this so that it can be fixed.

In the meantime, I have found that using another release of liquibase fixed my issue, and it has caused issues in other teams, so they changed for other versions of other teams, and how I'm having another issue (I could not deploy the fix to production, only to staging envs :man_facepalming: ) So I don't think I will be able to reproduce it easily, but anyway, I think a great way to address this issue and potentially other issues might be to improve the error handling so that it prints a bit more context in the stack trace, in particular the information you are asking for.


> Even better if you can find it and submit a PR to fix it.

Sorry, I'm really not proficient in Java, I can ask internally for somebody to look into it if it comes back but since we changed versions, I'm not sure it's going to come back. Feel free to close this if don't think there is anything to be done about the stack trace itself.

**Comment by molivasdat** [comment]
Hi @greg0ire Thanks for the feedback on being able to use another version that solves your problem.  We will add this to the list of issues to process to correct the NPE.

**Comment by MalloD12** [comment]
I've tried to reproduced this issue (by executing an `update` command) on below environment without success, I'm not sure either this issue has been already fixed on a newer version or I'm not testing it properly:

```
Liquibase Version: 4.15.0

Liquibase Integration & Version: CLI

Liquibase Extension(s) & Version: liquibase-percona 4.4.3

Database Vendor & Version: MySQL 8

Operating System Type & Version: Mac OS 12.1
```

@greg0ire Changelog is very useful, but could you please also share which command are you executing? So I can try to reproduce it in the same way.

**Comment by kevin-atx** [comment]
@greg0ire - we believe this issue has been fixed in Liquibase 4.15.0 and later. Please re-open this ticket if you are still seeing the problem.
